[SVG] Compute stroke-bounding-box lazily

Constellation opened the pull request:
#### 131951a5efc50b06d095a49525a5653e0fb86c67
<pre>
[SVG] Compute stroke-bounding-box lazily
<a href="https://bugs.webkit.org/show_bug.cgi?id=263065">https://bugs.webkit.org/show_bug.cgi?id=263065</a>
rdar://116852041

Reviewed by Cameron McCormack.

This is one step of the patch series implementing approximate repainting rect for SVG path, derived from blink&apos;s work[1].

This patch makes strokeBoundingBox computation lazy. updateShapeFromElement&apos;s responsibility gets reduced to,

    1. It needs to compute m_fillBoundingBox, but
    2. It *can* compute m_strokeBoundingBox if it is cheap

And if m_strokeBoundingBox is not computed in updateShapeFromElement, then we lazily compute it when strokeBoundingBox() function is called.
As a result of this responsibility reduction, LegacyRenderSVGShape::updateShapeFromElement becomes just creating a path and getting boundingBox.
So, this patch makes LegacyRenderSVGShape::updateShapeFromElement as pure virtual function, and we do this path materialization in each derived class&apos;
updateShapeFromElement. This makes it explicit that nobody is directly calling LegacyRenderSVGShape::updateShapeFromElement from layout function,
and it allows us to do sequence of clearing of related fields (m_shapeType, m_strokeBoundingBox etc.) done in each updateShapeFromElement function&apos;s prologue.

We also applied this to LSBE, and it requires many threading of RepaintRectCalculation. We add CalculateFastRepaintRect to SVGBoundingBoxComputation&apos;s option,
and thread RepaintRectCalculation information.

[1]: <a href="https://bugs.chromium.org/p/chromium/issues/detail?id=435097">https://bugs.chromium.org/p/chromium/issues/detail?id=435097</a>

* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::layoutChildren):
(WebCore::RenderSVGContainer::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGContainer.h:
* Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
(WebCore::RenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::checkIntersection):
(WebCore::RenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
(WebCore::RenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
(WebCore::RenderSVGPath::computeMarkerBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGPath.h:
* Source/WebCore/rendering/svg/RenderSVGRect.cpp:
(WebCore::RenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGResource.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
(WebCore::RenderSVGResourceMarker::markerBoundaries const):
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::calculateMaskContentRepaintRect):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::layoutChildren):
(WebCore::RenderSVGRoot::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::nodeAtPoint):
(WebCore::RenderSVGShape::strokeBoundingBox const):
(WebCore::RenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::RenderSVGShape::ensurePath):
(WebCore::RenderSVGShape::updateShapeFromElement): Deleted.
(WebCore::RenderSVGShape::calculateObjectBoundingBox const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
(WebCore::RenderSVGShape::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleShapeOrTextOrInline const):
(WebCore::SVGBoundingBoxComputation::handleRootOrContainer const):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::updateCachedBoundaries):
(WebCore::LegacyRenderSVGContainer::strokeBoundingBox const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
(WebCore::LegacyRenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::checkIntersection):
(WebCore::LegacyRenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
(WebCore::LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
(WebCore::LegacyRenderSVGPath::markerRect const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
(WebCore::LegacyRenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::calculateClipContentRepaintRect):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::updateCachedBoundaries):
(WebCore::LegacyRenderSVGRoot::strokeBoundingBox const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::strokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::ensurePath):
(WebCore::LegacyRenderSVGShape::updateShapeFromElement): Deleted.
(WebCore::LegacyRenderSVGShape::calculateObjectBoundingBox const): Deleted.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):

Canonical link: <a href="https://commits.webkit.org/269351@main">https://commits.webkit.org/269351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d20a3918dd57a622c292d99f433f51b2399f3347

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22327 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23402 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24227 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20663 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22577 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22856 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22259 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/19377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25082 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/19309 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/20232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/20463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20996 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17788 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20452 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24684 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2791 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20990 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
```diff
@@ -154,10 +154,27 @@ void LegacyRenderSVGContainer::addFocusRingRects(Vector<LayoutRect>& rects, cons
+    // https://bugs.webkit.org/show_bug.cgi?id=262409
+    if (!m_strokeBoundingBox)
+        m_strokeBoundingBox = SVGRenderSupport::computeStrokeBoundingBox(*this);
+    return m_strokeBoundingBox.value();
```
Nit: `*m_strokeBoundingBox` instead of `m_strokeBoundingBox.value()` is more common in WebCore. (Same in the other `strokeBoundingBox()` methods in other classes.)

Constellation replied:
Changed.

heycam commented on Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
```diff
@@ -50,12 +50,12 @@ void RenderSVGEllipse::updateShapeFromElement()
+    clearPath();
     m_shapeType = ShapeType::Empty;
+    m_strokeBoundingBox = std::nullopt;
     m_fillBoundingBox = FloatRect();
```
Very small nit: initializing in the other order is nicer, since that (a) matches the order of fields in the class, and (b) we often talk about "fill and stroke" of a shape in that order, not "stroke and fill".

Same in the other classes.

Constellation replied:
Changed.

heycam commented on Source/WebCore/rendering/svg/RenderSVGPath.cpp:
```diff
@@ -67,7 +67,7 @@ void RenderSVGPath::updateShapeFromElement()
 }
 
-FloatRect RenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps(FloatRect strokeBoundingBox) const
+FloatRect RenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps(RepaintRectCalculation, FloatRect strokeBoundingBox) const
```
Maybe I missed this in the previous review, but why does `RenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps` not include the marker rect like `LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps` does?

Constellation replied:
Yes. In new LBSE, stroke-bounding-box's meaning is different because SVG2's stroke-bounding-box is different from SVG1's stroke-bounding-box. So, SVG2 introduces decorated-bounding-box, and `computeMarkerBoundingBox` and its unite is done separately in `SVGBoundingBoxComputation::handleShapeOrTextOrInline`.

heycam replied:
I see, so it's because in LBSE we `RenderSVGShape::repaintRectInLocalCoordinates` calls `SVGBoundingBoxComputation::computeRepaintBoundingBox`, which includes markers through the functions you mention, rather than relying on a pre-computed repaint rect.

Constellation replied:
Yeah, and I think this is very good direction! We were using repaint bonding box for non-repainting purpose, and LBSE is avoiding it.

heycam commented on Source/WebCore/rendering/svg/SVGBoundingBoxComputation.h:
```diff
@@ -44,15 +44,16 @@ class SVGBoundingBoxComputation {
     static constexpr DecorationOptions strokeBoundingBoxDecoration = { DecorationOption::IncludeFillShape, DecorationOption::IncludeStrokeShape };
     static constexpr DecorationOptions filterBoundingBoxDecoration = { DecorationOption::OverrideBoxWithFilterBox, DecorationOption::OverrideBoxWithFilterBoxForChildren };
-    static constexpr DecorationOptions repaintBoundingBoxDecoration = { DecorationOption::IncludeFillShape, DecorationOption::IncludeStrokeShape, DecorationOption::IncludeMarkers, DecorationOption::IncludeClippers, DecorationOption::IncludeMaskers, DecorationOption::OverrideBoxWithFilterBox };
+    static constexpr DecorationOptions repaintBoundingBoxDecoration = { DecorationOption::IncludeFillShape, DecorationOption::IncludeStrokeShape, DecorationOption::IncludeMarkers, DecorationOption::IncludeClippers, DecorationOption::IncludeMaskers, DecorationOption::OverrideBoxWithFilterBox, DecorationOption::CalculateFastRepaintRect };
```
Oh, I like that the LBSE code makes this easy.

heycam commented on Source/WebCore/rendering/svg/RenderSVGPath.cpp:
```diff
@@ -262,7 +262,7 @@ FloatRect RenderSVGPath::computeMarkerBoundingBox(const SVGBoundingBoxComputatio
             // FIXME: [LBSE] Upstream RenderSVGResourceMarker changes
             // boundaries.unite(marker->computeMarkerBoundingBox(options, marker->markerTransformation(m_markerPositions[i].origin, m_markerPositions[i].angle, strokeWidth())));
-            boundaries.unite(marker->markerBoundaries(marker->markerTransformation(m_markerPositions[i].origin, m_markerPositions[i].angle, strokeWidth())));
+            boundaries.unite(marker->markerBoundaries(options.contains(SVGBoundingBoxComputation::DecorationOption::CalculateFastRepaintRect) ? RepaintRectCalculation::Fast : RepaintRectCalculation::Accurate, marker->markerTransformation(m_markerPositions[i].origin, m_markerPositions[i].angle, strokeWidth())));
```
This line is getting too long. Can you factor out the `RepaintRectCalculation` value computation into a local variable?

Constellation replied:
Changed.

heycam commented on Source/WebCore/rendering/svg/RenderSVGRect.cpp:
```diff
@@ -74,8 +75,8 @@ void RenderSVGRect::updateShapeFromElement()
     if (m_shapeType != ShapeType::Rectangle || hasNonScalingStroke()) {
-        // Fall back to RenderSVGShape
-        RenderSVGShape::updateShapeFromElement();
+        // Fallback to path-based approach if shape has a non-scaling stroke.
```
We're also using the fallback for rounded rectangles. I think "Fall back to path-based approach" is sufficient.

Constellation replied:
Oops, sounds good!

heycam commented on Source/WebCore/rendering/svg/RenderSVGResource.h:
```diff
@@ -43,6 +43,9 @@ enum class RenderSVGResourceMode {
     ApplyToText    = 1 << 2 // used in combination with ApplyTo{Fill|Stroke}Mode
 };
 
+
```
Nit: drop one of these blank lines.

Constellation replied:
Removed.

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
```diff
@@ -71,8 +72,8 @@ void LegacyRenderSVGRect::updateShapeFromElement()
     if (m_shapeType != ShapeType::Rectangle || hasNonScalingStroke()) {
-        // Fall back to LegacyRenderSVGShape
-        LegacyRenderSVGShape::updateShapeFromElement();
+        // Fallback to path-based approach if shape has a non-scaling stroke.
```
Same here.

Constellation replied:
Fixed.

heycam commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -192,7 +192,7 @@ FloatRect RenderSVGResourceMasker::resourceBoundingBox(const RenderObject& objec
 
     if (m_maskContentBoundaries.isEmpty())
-        calculateMaskContentRepaintRect();
+        calculateMaskContentRepaintRect(repaintRectCalculation);
```
I don't think this is right. `calculateMaskContentRepaintRect` will compute `m_maskContentBoundaries`, but we don't want that to be based on the `RepaintRectCalculation` value, since that might be different next time.

Constellation replied:
Oops, right. I'll change the code to keep both as cached.

heycam commented on Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
```diff
@@ -250,19 +259,20 @@ void SVGBoundingBoxComputation::adjustBoxForClippingAndEffects(const SVGBounding
                 if (auto* masker = resources->masker())
-                    box.intersect(masker->resourceBoundingBox(m_renderer));
+                    box.intersect(masker->resourceBoundingBox(m_renderer, repaintRectCalculation));
             }
```
My guess is that it's not that important to optimize the repaint rect calculations for these resources. I don't believe strokes of shapes are honored in `<clipPath>`, and I think it's rare for stroke shapes to be used inside a `<mask>`.

Unless you have evidence otherwise, given my comment above about `calculateMaskContentRepaintRect`, it might be better to have the definitions of these `resourceBoundingBox` functions always use accurate bounding box calculations.

Constellation replied:
OK! For now, I'll always use `RepaintRectCalculation::Accurate` for them!

heycam commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -166,14 +162,35 @@ void SVGRenderSupport::computeContainerBoundingBoxes(const RenderElement& contai
+
+FloatRect SVGRenderSupport::computeStrokeBoundingBox(const RenderElement& container)
+{
+    FloatRect strokeBoundingBox = FloatRect();
```
Maybe assert `container.isLegacySVGRoot() || container.isLegacySVGContainer()`?

Constellation replied:
Sounds good! Changed.

heycam reviewed (commented):
The general approach is good. Some specific questions and comments below.

Constellation reviewed (commented):
Thanks!

heycam commented on Source/WebCore/rendering/svg/SVGRenderSupport.h:
```diff
@@ -61,7 +61,8 @@ class SVGRenderSupport {
 
-    static void computeContainerBoundingBoxes(const RenderElement& container, FloatRect& objectBoundingBox, bool& objectBoundingBoxValid, FloatRect& strokeBoundingBox, FloatRect& repaintBoundingBox);
+    static void computeContainerBoundingBoxes(const RenderElement& container, FloatRect& objectBoundingBox, bool& objectBoundingBoxValid, FloatRect& repaintBoundingBox, RepaintRectCalculation = RepaintRectCalculation::Fast);
+    static FloatRect computeStrokeBoundingBox(const RenderElement& container);
```
Let's call this `computeContainerStrokeBoundingBox` for consistency with `computeContainerBoundingBoxes`.

Constellation replied:
Sounds good, changed!

Constellation reviewed (commented):
Thanks!

Constellation committed:
[SVG] Compute stroke-bounding-box lazily
https://bugs.webkit.org/show_bug.cgi?id=263065
rdar://116852041

Reviewed by Cameron McCormack.

This is one step of the patch series implementing approximate repainting rect for SVG path, derived from blink's work[1].

This patch makes strokeBoundingBox computation lazy. updateShapeFromElement's responsibility gets reduced to,

    1. It needs to compute m_fillBoundingBox, but
    2. It *can* compute m_strokeBoundingBox if it is cheap

And if m_strokeBoundingBox is not computed in updateShapeFromElement, then we lazily compute it when strokeBoundingBox() function is called.
As a result of this responsibility reduction, LegacyRenderSVGShape::updateShapeFromElement becomes just creating a path and getting boundingBox.
So, this patch makes LegacyRenderSVGShape::updateShapeFromElement as pure virtual function, and we do this path materialization in each derived class'
updateShapeFromElement. This makes it explicit that nobody is directly calling LegacyRenderSVGShape::updateShapeFromElement from layout function,
and it allows us to do sequence of clearing of related fields (m_shapeType, m_strokeBoundingBox etc.) done in each updateShapeFromElement function's prologue.

We also applied this to LSBE, and it requires many threading of RepaintRectCalculation. We add CalculateFastRepaintRect to SVGBoundingBoxComputation's option,
and thread RepaintRectCalculation information.

[1]: https://bugs.chromium.org/p/chromium/issues/detail?id=435097

* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::layoutChildren):
(WebCore::RenderSVGContainer::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGContainer.h:
* Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
(WebCore::RenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::checkIntersection):
(WebCore::RenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
(WebCore::RenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
(WebCore::RenderSVGPath::computeMarkerBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGPath.h:
* Source/WebCore/rendering/svg/RenderSVGRect.cpp:
(WebCore::RenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGResource.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
(WebCore::RenderSVGResourceMarker::markerBoundaries const):
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::calculateMaskContentRepaintRect):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::layoutChildren):
(WebCore::RenderSVGRoot::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::nodeAtPoint):
(WebCore::RenderSVGShape::strokeBoundingBox const):
(WebCore::RenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::RenderSVGShape::ensurePath):
(WebCore::RenderSVGShape::updateShapeFromElement): Deleted.
(WebCore::RenderSVGShape::calculateObjectBoundingBox const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
(WebCore::RenderSVGShape::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleShapeOrTextOrInline const):
(WebCore::SVGBoundingBoxComputation::handleRootOrContainer const):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::updateCachedBoundaries):
(WebCore::LegacyRenderSVGContainer::strokeBoundingBox const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
(WebCore::LegacyRenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::checkIntersection):
(WebCore::LegacyRenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
(WebCore::LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
(WebCore::LegacyRenderSVGPath::markerRect const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
(WebCore::LegacyRenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::calculateClipContentRepaintRect):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::updateCachedBoundaries):
(WebCore::LegacyRenderSVGRoot::strokeBoundingBox const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::strokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::ensurePath):
(WebCore::LegacyRenderSVGShape::updateShapeFromElement): Deleted.
(WebCore::LegacyRenderSVGShape::calculateObjectBoundingBox const): Deleted.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):

Canonical link: https://commits.webkit.org/269351@main
